README.md
<div align="center">

# feedback-js

[![Build](https://github.com/BetaHuhn/feedback-js/workflows/Build/badge.svg)](https://github.com/BetaHuhn/feedback-js/actions?query=workflow%3ABuild) [![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/BetaHuhn/feedback-js/blob/master/LICENSE) [![npm](https://img.shields.io/npm/v/@betahuhn/feedback-js)](https://www.npmjs.com/package/@betahuhn/feedback-js) [![npm bundle size](https://img.badgesize.io/betahuhn/feedback-js/master/dist/feedback-js.min.js?compression=gzip)](https://github.com/BetaHuhn/feedback-js)

Simple feedback modal for any website

![preview](https://cdn.mxis.ch/assets/feedback-js/preview.gif)

[🔮 Live Demo](https://codepen.io/BetaHuhn/pen/MWebeXp)
<br/>

</div>

## ⚡ Features

- Easy to integrate with any site (via [script tag](#script-tag) or [NPM](#npm))
- Simple and modern design
- Customize feedback types and text
- Works with any backend (self-hosted or form providers)
- Use the default feedback button or trigger the modal from any HTML element

## 🚀 Get started

### Via CDN

Add this to your HTML page:

```html
<script src="https://cdn.jsdelivr.net/npm/@betahuhn/feedback-js/dist/feedback-js.min.js" data-feedback-endpoint="https://your-custom-endpoint"></script>
```

### Via NPM

Install [feedback-js](https://github.com/BetaHuhn/feedback-js) using NPM:

```sh
npm install @betahuhn/feedback-js
```

Then add the following JavaScript code:

```javascript
import Feedback from '@betahuhn/feedback-js'

new Feedback({ endpoint: 'https://example.com/feedback' }).renderButton()
```

By default [feedback-js](https://github.com/BetaHuhn/feedback-js) will add a feedback button to the bottom right corner of your page with the default colors and text. This can be configured using the [options](https://github.com/BetaHuhn/feedback-js#options) object.

## 📚 Setup

There are multiple ways to use [feedback-js](https://github.com/BetaHuhn/feedback-js). The easiest is to include the provided script tag in your HTML file and specify the form endpoint with `data-feedback-endpoint`:

```html
<script src="https://cdn.jsdelivr.net/npm/@betahuhn/feedback-js/dist/feedback-js.min.js" data-feedback-endpoint="https://your-custom-endpoint"></script>
```

If you want to customize [feedback-js](https://github.com/BetaHuhn/feedback-js) further, you can use the `data-feedback-opts` attribute:

```html
<script data-feedback-opts='{ "endpoint": "http://172.21.52.196:6600/form/feedback", "id": "example", "emailField": true }'></script>
```

You can add the attributes to any HTML element, not just the `script` tag e.g. `<div data-feedback-opts=""></div>`. 

You can also set it up manually with JavaScript:

```javascript
import Feedback from '@betahuhn/feedback-js'

const options = {
    endpoint: 'https://example.com/feedback',
    id: 'example',
    emailField: true
}

new Feedback(options).renderButton()
```

They both achieve the same result.

### Feedback Button

By default if you specify a `data-feedback-endpoint` or `data-feedback-opts` attribute or call `.renderButton()` the default feedback button will be rendered. 

You can also trigger the modal from any element by adding `data-feedback-trigger` to it:

```html
<button data-feedback-trigger>Give Feedback</button>
```

This will open the modal and the user can give his feedback.

### Backend

Once the user submits the form, the data will be sent to the specified endpoint. 

You will have to handle the submission on the backend yourself. [feedback-js](https://github.com/BetaHuhn/feedback-js) will make a POST request to your specified endpoint with the following body:

```json
{
    "id": "example",
    "email": "hello@mxis.ch",
    "feedbackType": "issue",
    "url": "https://example.com",
    "message": "When I click x nothing happens."
}
```

<details>
 <summary>See Node.js example</summary>
	
```js
const express = require('express')
const app = express()
const port = 3000

app.post('/feedback', async (req, res) => {
	const { id, feedbackType, message, email, url } = req.body

	console.log(`New ${ feedbackType } feedback for form ${ id } from user ${ email } on page ${ url }: ${ message }`)
	// do something with feedback

	res.send('ok')
})

app.listen(port, () => {
	console.log(`Listening at http://localhost:${ port }`)
})
```
	
</details>

## ⚙️ Options

You can customize [feedback-js](https://github.com/BetaHuhn/feedback-js) by passing a options object to `new Feedback()` or use the `data-feedback-opts` attribute.

<details>
<summary>All available options:</summary>

```js
const options = {
    id: 'feedback', // id to identify the form on the backend
    endpoint: 'https://example.com/feedback', // enpoint of your backend to handle the submission
    emailField: true, // show email input field, default: false
    events: false, // Emit feedback-submit event instead of sending data to endpoint
    forceShowButton: false, // Show the default button even if you use `data-feedback-trigger`
    types: { // Feedback types
        general: {
            text: 'General Feedback',
            icon: '😁'
        },
        idea: {
            text: 'I have an idea',
            icon: '💡'
        },
        bug: {
            text: 'I found an issue',
            icon: '🐞'
        }
    },
    btnTitle: 'Feedback', // title of button
    title: 'Company Feedback', // text at the top
    inputPlaceholder: 'Your feedback goes here!',
    emailPlaceholder: 'Email address (optional)',
    submitText: 'Submit', // text for submit button
	backText: 'Back', // text for back button
    contactText: 'Or send an email!', // text for other contact option
    contactLink: 'mailto:hello@mxis.ch', // link for other contact option
    typeMessage: 'What feedback do you have?', // message for selecting feedback type
    success: 'Thanks! 👊', // message displayed on successfull submission
    failedTitle: 'Oops, an error ocurred!', // title displayed on error
    failedMessage: 'Please try again. If this keeps happening, try to send an email instead.', // default error message if backend doesn't return one
    position: 'right', // position of button left/right
    primary: 'rgb(53, 222, 118)', // primary color
    background: '#fff', // background color
    color: '#000' // font color
}

const feedback = new Feedback(options)
feedback.renderButton()
```

</details>

## Email field

By default the email field will be hidden, you can enable it with the `emailField` option.

### Custom submission method

If you want to handle the form submission yourself, you can enable `events` and listen for the `feedback-submit` event:

```js
import Feedback from '@betahuhn/feedback-js'

const options = {
    events: true
}

new Feedback(options).renderButton()

window.addEventListener('feedback-submit', (event) => {
    console.log(event.detail) // will contain the feedback data which would have been sent to the endpoint
})
```

### Custom feedback types

By default there are 3 feedback types, general, idea and bug. You can modify their text/icons or even add your own with the `types` option:

```js
import Feedback from '@betahuhn/feedback-js'

const options = {
    endpoint: 'https://example.com/feedback',
    types: {
        general: {
            text: 'General Feedback',
            icon: '⚠️'
        },
        idea: {
            text: 'I have an idea',
            icon: '💡'
        },
        love: {
            text: 'Send love',
            icon: '💖'
        }
    }
}

new Feedback(options).renderButton()
```

## 💻 Development

Issues and PRs are very welcome!

The actual source code of this library is in the `feedback.js` file in the `src` folder.

- run `yarn lint` or `npm run lint` to run eslint.
- run `yarn watch` or `npm run watch` to watch for changes and build to the `dist` folder.
- run `yarn build` or `npm run build` to produce a production version of [feedback-js](https://github.com/BetaHuhn/feedback-js) in the `dist` folder.

## ❔ About

This library was developed by me ([@betahuhn](https://github.com/BetaHuhn)) in my free time. If you want to support me:

[![Donate via PayPal](https://img.shields.io/badge/paypal-donate-009cde.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=394RTSBEEEFEE)

### Credits

The design of the feedback form was inspired by [@kangabru's](https://github.com/kangabru/) feedback form on the [Panda Snap](https://pandasnap.io/) dashboard.

### License

Copyright 2021 Maximilian Schiller

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details


CHANGELOG.md
## [v2.1.25] - 2023-10-30

[Release notes](https://github.com/BetaHuhn/feedback-js/releases/tag/v2.1.25) · [Compare](https://github.com/BetaHuhn/feedback-js/compare/v2.1.24...v2.1.25) · [Tag](https://github.com/BetaHuhn/feedback-js/tree/v2.1.25) · Archive ([zip](https://github.com/BetaHuhn/feedback-js/archive/v2.1.25.zip) · [tar.gz](https://github.com/BetaHuhn/feedback-js/archive/v2.1.25.tar.gz))

### Dependency updates

- [`bc893c8`](https://github.com/BetaHuhn/feedback-js/commit/bc893c8)  Bump browserify-sign from 4.2.1 to 4.2.2

## [v2.1.24] - 2023-10-23

[Release notes](https://github.com/BetaHuhn/feedback-js/releases/tag/v2.1.24) · [Compare](https://github.com/BetaHuhn/feedback-js/compare/v2.1.23...v2.1.24) · [Tag](https://github.com/BetaHuhn/feedback-js/tree/v2.1.24) · Archive ([zip](https://github.com/BetaHuhn/feedback-js/archive/v2.1.24.zip) · [tar.gz](https://github.com/BetaHuhn/feedback-js/archive/v2.1.24.tar.gz))

### Dependency updates

- [`06455a7`](https://github.com/BetaHuhn/feedback-js/commit/06455a7)  Bump @babel/traverse from 7.12.12 to 7.23.2

## [v2.1.23] - 2023-07-24

[Release notes](https://github.com/BetaHuhn/feedback-js/releases/tag/v2.1.23) · [Compare](https://github.com/BetaHuhn/feedback-js/compare/v2.1.22...v2.1.23) · [Tag](https://github.com/BetaHuhn/feedback-js/tree/v2.1.23) · Archive ([zip](https://github.com/BetaHuhn/feedback-js/archive/v2.1.23.zip) · [tar.gz](https://github.com/BetaHuhn/feedback-js/archive/v2.1.23.tar.gz))

### Dependency updates

- [`555170a`](https://github.com/BetaHuhn/feedback-js/commit/555170a)  Bump word-wrap from 1.2.3 to 1.2.4

## [v2.1.22] - 2023-02-06

[Release notes](https://github.com/BetaHuhn/feedback-js/releases/tag/v2.1.22) · [Compare](https://github.com/BetaHuhn/feedback-js/compare/v2.1.21...v2.1.22) · [Tag](https://github.com/BetaHuhn/feedback-js/tree/v2.1.22) · Archive ([zip](https://github.com/BetaHuhn/feedback-js/archive/v2.1.22.zip) · [tar.gz](https://github.com/BetaHuhn/feedback-js/archive/v2.1.22.tar.gz))

### Dependency updates

- [`eb3041b`](https://github.com/BetaHuhn/feedback-js/commit/eb3041b)  Bump http-cache-semantics from 4.1.0 to 4.1.1

## [v2.1.21] - 2023-01-09

[Release notes](https://github.com/BetaHuhn/feedback-js/releases/tag/v2.1.21) · [Compare](https://github.com/BetaHuhn/feedback-js/compare/v2.1.20...v2.1.21) · [Tag](https://github.com/BetaHuhn/feedback-js/tree/v2.1.21) · Archive ([zip](https://github.com/BetaHuhn/feedback-js/archive/v2.1.21.zip) · [tar.gz](https://github.com/BetaHuhn/feedback-js/archive/v2.1.21.tar.gz))

### Dependency updates

- [`e3a0ba7`](https://github.com/BetaHuhn/feedback-js/commit/e3a0ba7)  Bump json5 from 2.1.3 to 2.2.3

## [v2.1.20] - 2022-12-12

[Release notes](https://github.com/BetaHuhn/feedback-js/releases/tag/v2.1.20) · [Compare](https://github.com/BetaHuhn/feedback-js/compare/v2.1.19...v2.1.20) · [Tag](https://github.com/BetaHuhn/feedback-js/tree/v2.1.20) · Archive ([zip](https://github.com/BetaHuhn/feedback-js/archive/v2.1.20.zip) · [tar.gz](https://github.com/BetaHuhn/feedback-js/archive/v2.1.20.tar.gz))

### Dependency updates

- [`00c9aa6`](https://github.com/BetaHuhn/feedback-js/commit/00c9aa6)  Bump qs from 6.5.2 to 6.5.3

## [v2.1.19] - 2022-12-05

[Release notes](https://github.com/BetaHuhn/feedback-js/releases/tag/v2.1.19) · [Compare](https://github.com/BetaHuhn/feedback-js/compare/v2.1.18...v2.1.19) · [Tag](https://github.com/BetaHuhn/feedback-js/tree/v2.1.19) · Archive ([zip](https://github.com/BetaHuhn/feedback-js/archive/v2.1.19.zip) · [tar.gz](https://github.com/BetaHuhn/feedback-js/archive/v2.1.19.tar.gz))

### Dependency updates

- [`1ade791`](https://github.com/BetaHuhn/feedback-js/commit/1ade791)  Bump stefanzweifel/git-auto-commit-action
- [`1b83665`](https://github.com/BetaHuhn/feedback-js/commit/1b83665)  Bump minimatch from 3.0.4 to 3.1.2
- [`e61429a`](https://github.com/BetaHuhn/feedback-js/commit/e61429a)  Bump decode-uri-component from 0.2.0 to 0.2.2

## [v2.1.18] - 2022-11-14

[Release notes](https://github.com/BetaHuhn/feedback-js/releases/tag/v2.1.18) · [Compare](https://github.com/BetaHuhn/feedback-js/compare/v2.1.17...v2.1.18) · [Tag](https://github.com/BetaHuhn/feedback-js/tree/v2.1.18) · Archive ([zip](https://github.com/BetaHuhn/feedback-js/archive/v2.1.18.zip) · [tar.gz](https://github.com/BetaHuhn/feedback-js/archive/v2.1.18.tar.gz))

### Dependency updates

- [`8925ca5`](https://github.com/BetaHuhn/feedback-js/commit/8925ca5)  Bump stefanzweifel/git-auto-commit-action

## [v2.1.17] - 2022-10-31

[Release notes](https://github.com/BetaHuhn/feedback-js/releases/tag/v2.1.17) · [Compare](https://github.com/BetaHuhn/feedback-js/compare/v2.1.16...v2.1.17) · [Tag](https://github.com/BetaHuhn/feedback-js/tree/v2.1.17) · Archive ([zip](https://github.com/BetaHuhn/feedback-js/archive/v2.1.17.zip) · [tar.gz](https://github.com/BetaHuhn/feedback-js/archive/v2.1.17.tar.gz))

### Dependency updates

- [`c0bb510`](https://github.com/BetaHuhn/feedback-js/commit/c0bb510)  Bump stefanzweifel/git-auto-commit-action
- [`5cbab97`](https://github.com/BetaHuhn/feedback-js/commit/5cbab97)  Bump stefanzweifel/git-auto-commit-action

## [v2.1.16] - 2022-10-17

[Release notes](https://github.com/BetaHuhn/feedback-js/releases/tag/v2.1.16) · [Compare](https://github.com/BetaHuhn/feedback-js/compare/v2.1.15...v2.1.16) · [Tag](https://github.com/BetaHuhn/feedback-js/tree/v2.1.16) · Archive ([zip](https://github.com/BetaHuhn/feedback-js/archive/v2.1.16.zip) · [tar.gz](https://github.com/BetaHuhn/feedback-js/archive/v2.1.16.tar.gz))

### Dependency updates

- [`ac746ec`](https://github.com/BetaHuhn/feedback-js/commit/ac746ec)  Bump stefanzweifel/git-auto-commit-action

## [v2.1.15] - 2022-10-03


src/index.js
import Feedback, { IS_BROWSER } from './feedback'
export default Feedback

const detect = () => {
	const optsElem = document.querySelector('[data-feedback-opts]')
	const endpointElem = document.querySelector('[data-feedback-endpoint]')
	const buttonElems = document.querySelectorAll('[data-feedback-trigger]')

	// If no attributes are found, assume programmatic usage and attach Feedback class to window
	if (!optsElem && !endpointElem && buttonElems.length < 1) {
		window.Feedback = Feedback

		return
	}

	// Parse options specified in data-feedback-opts and data-feedback-endpoint attributes
	const attributeValue = optsElem && optsElem.getAttribute('data-feedback-opts') || '{}'
	const endpointValue = endpointElem && { endpoint: endpointElem.getAttribute('data-feedback-endpoint') } || {}
	const options = Object.assign({}, JSON.parse(attributeValue), endpointValue)

	// Initialize the widget and attach styles
	window.feedback = new Feedback(options)

	// Render the default button
	const renderDefaultButton = buttonElems.length < 1
	if (renderDefaultButton || window.feedback.options.forceShowButton) {
		window.feedback.renderButton()
	}

	// Attach event listeners to data-drkmd-toggle elements
	if (!renderDefaultButton) {
		buttonElems.forEach((item) => {
			item.addEventListener('click', () => {
				window.feedback.renderModal()
			})
		})
	}
}

if (IS_BROWSER) {
	window.addEventListener('DOMContentLoaded', () => {
		detect()
	})
} else {
	console.warn('[feedback-js] Detected environment without a `window` object')
}

src/feedback.js
export const IS_BROWSER = typeof window !== 'undefined'

export default class Feedback {
	/**
	 * Feedback item
	 * @param {object} options - object containing feedback options
	 */
	constructor(options) {
		if (!IS_BROWSER) {
			console.warn('Detected environment without a `window` object')
			return
		}

		const defaultOptions = {
			id: 'feedback',
			endpoint: '',
			events: false,
			emailField: false,
			forceShowButton: false,
			btnTitle: 'Feedback',
			title: 'Feedback',
			contactText: 'Want to chat?',
			contactLink: '',
			typeMessage: 'What feedback do you have?',
			success: 'Thanks! 👊',
			inputPlaceholder: 'Your feedback goes here!',
			emailPlaceholder: 'Email address (optional)',
			submitText: 'Submit',
			backText: 'Back',
			failedTitle: 'Oops, an error ocurred!',
			failedMessage: 'Please try again. If this keeps happening, try to send an email instead.',
			position: 'right',
			primary: 'rgb(53, 222, 118)',
			background: '#fff',
			color: '#000',
			types: {
				general: {
					text: 'General Feedback',
					icon: '😁'
				},
				idea: {
					text: 'I have an idea',
					icon: '💡'
				},
				bug: {
					text: 'I found an issue',
					icon: '🐞'
				}
			}
		}

		// Parse the provided options and merge with defaults
		this.options = Object.assign({}, defaultOptions, options)

		// Create a new root element for feedback-js to place it's elements in
		const div = document.createElement('div')
		div.id = 'feedback-root'
		document.body.insertBefore(div, document.body.firstChild)
		this.root = div

		// Add a comment to the dom
		const comment = document.createComment('feedback-js modal code')
		document.body.insertBefore(comment, document.body.firstChild)

		// Add the required styles to the page
		this._addStyle()
	}

	/**
	 * Render the default button
	 */
	renderButton() {
		if (!this.root) return

		this.showDefaultBtn = true

		const html = `
			<div class="feedback-btn-wrapper">
				<button id="feedback-btn" title="Give feedback">
					<svg class="inline w-5 h-5" viewBox="0 0 20 20" fill="currentColor"><path fill-rule="evenodd" d="M18 13V5a2 2 0 00-2-2H4a2 2 0 00-2 2v8a2 2 0 002 2h3l3 3 3-3h3a2 2 0 002-2zM5 7a1 1 0 011-1h8a1 1 0 110 2H6a1 1 0 01-1-1zm1 3a1 1 0 100 2h3a1 1 0 100-2H6z" clip-rule="evenodd"></path></svg>
					<span>${ this.options.btnTitle }</span>
				</button>
			</div>
		`

		this.root.innerHTML = html

		const button = document.getElementById('feedback-btn')
		button.addEventListener('click', () => {
			this.renderModal()
		})
	}

	/**
	 * Render the feedback modal
	 */
	renderModal() {
		if (!this.root) return

		const html = `
			<div class="feedback-wrapper">
				<div class="feedback-main">
					<div class="feedback-header">
						<p>${ this.options.title }</p>
						${ this.options.contactLink.length > 0 ? '<a href=' + this.options.contactLink + '>' + this.options.contactText + '</a>' : '' }
					</div>
					<div class="feedback-content">
						<p>${ this.options.typeMessage }</p>
						<div class="feedback-content-list">
							${ Object.entries(this.options.types).reduce((prev, [ id, item ]) => prev += `<button id="feedback-item-${ id }" class="feedback-item"><span>${ item.icon }</span>${ item.text }</button>`, '') }
						</div>
					</div>
				</div>
				<div class="feedback-close">
					<button id="feedback-close">
						<svg class="w-5 h-5" fill="none" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" viewBox="0 0 24 24" stroke="currentColor"><path d="M6 18L18 6M6 6l12 12"></path></svg>
					</button>
				</div>
			</div>
		`

		this.root.innerHTML = html

		const button = document.getElementById('feedback-close')
		button.addEventListener('click', () => {
			this.closeModal()
		})

		Object.keys(this.options.types).forEach((id) => {
			const elem = document.getElementById(`feedback-item-${ id }`)

			elem.onclick = () => {
				this.renderForm(id)
			}
		})
	}

	/**
	 * Render the form for the given feedback type.
	 *
	 * The type needs to be specified in the `types` option.
	 * @param {String} type Feedback Type
	 */
	renderForm(type) {
		if (!this.root) return

		const feedbackType = this.options.types[type]

		const html = `
			<div class="feedback-wrapper">
				<div class="feedback-main">
					<div class="feedback-header">
						<p>${ feedbackType.icon } ${ feedbackType.text }</p>
					</div>
					<div class="feedback-content">
							${ this.options.emailField ? `<input id="feedback-email" type="email" name="email" placeholder="${ this.options.emailPlaceholder }">` : '' }
							<textarea id="feedback-message" name="feedback" autofocus type="text" maxlength="500" rows="5" placeholder="${ this.options.inputPlaceholder }"></textarea>
							<div id="feedback-actions" class="feedback-actions">
								<button type="button" id="feedback-back">${ this.options.backText }</button>
								<button type="submit" id="feedback-submit">${ this.options.submitText }</button>
							</div>
					</div>
				</div>
				<div class="feedback-close">
					<button id="feedback-close">
						<svg class="w-5 h-5" fill="none" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" viewBox="0 0 24 24" stroke="currentColor"><path d="M6 18L18 6M6 6l12 12"></path></svg>
					</button>
				</div>
			</div>
		`

		this.current = type
		this.root.innerHTML = html

		document.getElementById('feedback-message').focus()

		const button = document.getElementById('feedback-close')
		button.addEventListener('click', () => {
			this.closeModal()
		})

		const back = document.getElementById('feedback-back')
		back.addEventListener('click', () => {
			this.renderModal()
		})

		const submit = document.getElementById('feedback-submit')
		submit.addEventListener('click', () => {
			this.submitForm()
		})
	}

	/**
	 * Close/hide the feedback modal.
	 *
	 * If the default button was rendered before, it will be shown again.
	 */
	closeModal() {
		this.root.innerHTML = ''

		if (this.showDefaultBtn) {
			this.renderButton()
		}
	}

	/**
	 * Submit the form.
	 *
	 * Will be called when the user clicks submit.
	 */
	submitForm() {
		const message = document.getElementById('feedback-message').value
		const email = this.options.emailField ? document.getElementById('feedback-email').value : undefined

		const data = {
			id: this.options.id,
			email: email,
			feedbackType: this.current,
			url: window.location.href,
			message: message
		}

		if (this.options.events) {
			const event = new CustomEvent('feedback-submit', { detail: data })
			window.dispatchEvent(event)
			this.renderSuccess()
			return
		}

		this.sendToEndpoint(data)
	}

	/**
	 * Send the given feedback to the specified endpoint.
	 * @param {Object} data - the feedback data
	 */
	sendToEndpoint(data) {
		this.renderLoading()

		const request = new XMLHttpRequest()
		request.open('POST', this.options.endpoint)
		request.setRequestHeader('Content-type', 'application/json')
		request.send(JSON.stringify(data))
		request.onreadystatechange = () => {
			if (request.readyState === 4) {
				if (request.status === 200) {
					return this.renderSuccess()
				}

				this.renderFailed()
			}
		}
	}

	/**
	 * Render the loading state.
	 */
	renderLoading() {
		if (!this.root) return

		const html = `
			<button id="feedback-loading"><div class="feedback-loader"><div></div><div></div><div></div><div></div></div>Loading</button>
		`

		document.getElementById('feedback-actions').innerHTML = html

		const button = document.getElementById('feedback-close')
		button.addEventListener('click', () => {
			this.closeModal()
		})
	}

	/**
	 * Render the success state.
	 */
	renderSuccess() {
		if (!this.root) return

		const html = `
			<div class="feedback-btn-wrapper">
				<button id="feedback-btn" title="Give feedback">
					<span>${ this.options.success }</span>
				</button>
			</div>
		`

		this.root.innerHTML = html

		setTimeout(() => {
			this.renderButton()
		}, 3000)
	}

	/**
	 * Render the failed state.
	 */
	renderFailed() {
		if (!this.root) return

		const html = `
			<div class="feedback-wrapper">
				<div class="feedback-main">
					<div class="feedback-header">
						<p>${ this.options.failedTitle }</p>
					</div>
					<div class="feedback-content">
						<p>${ this.options.failedMessage }</p>
					</div>
				</div>
				<div class="feedback-close">
					<button id="feedback-close">
						<svg class="w-5 h-5" fill="none" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" viewBox="0 0 24 24" stroke="currentColor"><path d="M6 18L18 6M6 6l12 12"></path></svg>
					</button>
				</div>
			</div>
		`

		this.root.innerHTML = html

		const button = document.getElementById('feedback-close')
		button.addEventListener('click', () => {
			this.closeModal()
		})
	}

	_addStyle() {
		const css = `
			#feedback-root{
				font-family: system-ui,-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,Noto Sans,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol,Noto Color Emoji;
			}

			.feedback-wrapper{
				position: fixed;
				z-index: 1000;
				bottom: 0;
				${ this.options.position === 'left' ? 'left: 0' : 'right: 0' };
				margin: 2rem;
				width: 100%;
				max-width: 20rem;
			}

			.feedback-main{
				background-color: ${ this.options.background };
				color: ${ this.options.color };
				border-radius: 0.6rem;
				text-align: center;
				overflow: hidden;
				box-shadow: 0 10px 15px -3px rgba(0,0,0,.1), 0 4px 6px -2px rgba(0,0,0,.05);
			}

			.feedback-header{
				color: ${ this.options.background };
				background-color: ${ this.options.primary };
				padding: 0.8rem 1.25rem;
			}

			.feedback-header p{
				margin: 0;
				font-weight: 600;
				font-size: 1.2rem;
			}

			.feedback-header a{
				margin: 0;
				outline: 0;
				font-size: 1rem;
				margin-top: 0.5rem;
				display: block;
				color: ${ this.options.background };
				border: 2px solid ${ this.options.primary };
				text-decoration: none;
				padding: 5px 10px
			}

			.feedback-header a:hover{
				text-decoration: underline;
			}

			.feedback-header a:focus{
				border: 2px solid ${ this.options.background };
				border-radius: 5px;
			}

			.feedback-content{
				padding: 0.7rem 1.5rem;
				display: flex;
				align-items: center;
				flex-direction: column;
			}

			.feedback-content p{
				margin: 0;
				font-size: 1rem;
				font-weight: 600;
				margin-top: 0.5rem;
				margin-bottom: 0.6rem;
			}

			.feedback-content input{
				border: 3px solid ${ this.options.background };
				filter: brightness(95%);
				border-radius: 10px;
				outline: 0;
				padding: 10px;
				margin-bottom: 0.5rem;
				width: 100%;
				box-sizing: border-box;
				font-size: 1rem;
				font-family: system-ui,-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,Noto Sans,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol,Noto Color Emoji;
			}

			.feedback-content textarea{
				overflow: auto;
				border: 3px solid ${ this.options.background };
				filter: brightness(95%);
				border-radius: 10px;
				outline: 0;
				padding: 10px;
				width: 100%;
				box-sizing: border-box;
				resize: none;
				font-size: 1rem;
				font-family: system-ui,-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,Noto Sans,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol,Noto Color Emoji;
			}

			.feedback-content textarea:focus,
			.feedback-content input:focus{
				border: 3px solid ${ this.options.primary };
			}

			.feedback-actions{
				display: flex;
				margin-top: 1rem;
				width: 100%;
			}

			.feedback-actions button{
				padding: 0.5rem;
				border-radius: 10px;
				cursor: pointer;
				margin: 0;
				outline: 0;
				border: 3px solid rgba(255, 255, 255, 0);
				appearance: none;
				font-family: system-ui,-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,Noto Sans,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol,Noto Color Emoji;
				user-select: none;
				box-shadow: 0 1px 3px 0 rgba(0,0,0,.1);
			}

			.feedback-actions button:focus {
				border: 3px solid ${ this.options.primary };
				filter: brightness(105%);
			}
			
			.feedback-actions button:active {
				transform: scale(0.95);
			}

			#feedback-back{
				background: ${ this.options.background };
				color: rgba(58,71,65);
			}

			#feedback-submit{
				margin-left: auto;
				background: ${ this.options.primary };
				color: ${ this.options.background };
				font-weight: 700;
			}

			#feedback-loading{
				margin-left: auto;
				background: ${ this.options.primary };
				color: ${ this.options.background };
				font-weight: 700;
			}

			.feedback-loader{
				display: inline-block;
				position: relative;
				width: 15px;
				height: 15px;
				margin-right: 1rem;
			}

			.feedback-loader div {
				box-sizing: border-box;
				display: block;
				position: absolute;
				width: 20px;
				height: 20px;
				margin: 0;
				border: 3px solid ${ this.options.background };
				border-radius: 50%;
				animation: feedback-loader 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
				border-color: ${ this.options.background } transparent transparent transparent;
			}

			.feedback-loader div:nth-child(1) {
				animation-delay: -0.45s;
			}

			.feedback-loader div:nth-child(2) {
				animation-delay: -0.3s;
			}

			.feedback-loader div:nth-child(3) {
				animation-delay: -0.15s;
			}

			@keyframes feedback-loader {
				0% {
					transform: rotate(0deg);
				}
				100% {
					transform: rotate(360deg);
				}
			}
			  
			.feedback-content-list{
				display: flex;
				flex-direction: column;
				align-items: flex-start;
				margin-left: auto;
				margin-right: auto;
				margin-bottom: 1rem;
			}

			.feedback-item {
				width: 100%;
				cursor: pointer;
				user-select:none;
				border:0;
				outline: 0;
				background: 0;
				font-size: 0.9rem;
				text-align: left;
				display: flex;
				align-items: center;
				flex-direction: row;
				color: ${ this.options.color };
				background-color: ${ this.options.background };
				border: 2px solid ${ this.options.background };
				border-radius: 10px;
				padding: 10px;
			}

			.feedback-item:hover{
				filter: brightness(95%);
			}

			.feedback-item:focus{
				border: 2px solid ${ this.options.primary };
				border-radius: 5px;
			}

			.feedback-item span{
				width: 1.5rem;
				margin-right: 0.2rem;
				display: block;
				text-align: center;
			}

			.feedback-close{
				margin-top: .75rem;
				display: flex;
				align-items: center;
				justify-content: ${ this.options.position === 'left' ? 'flex-start' : 'flex-end' };
				flex-direction: row;
			}

			.feedback-close button{
				box-shadow: 0 10px 15px -3px rgba(0,0,0,.1), 0 4px 6px -2px rgba(0,0,0,.05);
				cursor: pointer;
				user-select:none;
				border:0;
				outline: 0;
				border-radius: 9999px;
				padding: .5rem;
				background: ${ this.options.background };
				color: ${ this.options.color };
				border: 2px solid ${ this.options.background };
			}

			.feedback-close button:focus{
				border: 2px solid ${ this.options.primary };
			}

			.feedback-close button:hover{
				filter: brightness(95%);
			}

			.feedback-close button:active {
				transform: scale(0.95);
			}

			.feedback-close svg{
				display: block;
				width: 1.25rem;
				height: 1.25rem;
			}

			.feedback-btn-wrapper{
				position: fixed;
				z-index: 1000;
				bottom: 0;
				${ this.options.position === 'left' ? 'left: 0' : 'right: 0' };
				margin: 2rem;
			}

			#feedback-btn{
				height: 30px;
				display: flex;
				align-items: center;
				padding: 1.2rem 0.7rem;
				border-radius: 0.4rem;
				cursor: pointer;
				box-shadow: 0 10px 15px -3px rgba(0,0,0,.1), 0 4px 6px -2px rgba(0,0,0,.05);
				user-select:none;
				border:0;
				outline: 0;
				color: ${ this.options.background };
				background-color: ${ this.options.primary };
				transition: filter .4s ease;
			}

			#feedback-btn:hover, #feedback-btn:focus{
				filter: brightness(105%);
			}

			#feedback-btn:active {
				transform: scale(0.95);
			}

			#feedback-btn svg{
				width: 1.25rem;
				height: 1.25rem;
			}

			#feedback-btn span{
				font-weight: 700;
				font-size: 1rem;
				margin-left: .5rem;
			}
		`

		const comment = document.createComment('feedback-js stylesheet')
		document.head.appendChild(comment)

		const styleElement = document.createElement("style")
		styleElement.innerHTML = css  
		document.head.appendChild(styleElement)
	}
}
